### PR TITLE
Feature/more documentation

### PR DESCRIPTION
**Pull request by lightsprint09**
Fixes issues: 

New feature:

#### Make sure to check all boxes before merging

- [x] Method/Class documentation
- [x] README.md documentation
- [x] Unit tests for new features/regressions

### CODE REVIEW

File: Source/BasicNetworkService.swift [hunk 1]
```diff
@@ -46,6 +54,31 @@ public final class BasicNetworkService: NetworkService {
...
+    /**
+     Fetches a resource asynchronously from remote location. Execution of the requests starts immediately.
+     Execution happens on no specific queue. It dependes on the network access which queue is used.
+     Once execution is finished either the completion block or the error block gets called.
+     You can decide on which queue these blocks get called.
+     
+     **Example**:
+     ```swift
+     let networkService: NetworkService = //
+     let resource: Ressource<String> = //
```
Comment (IsabellSchmelzeisen) [inline Source/BasicNetworkService.swift]: Ressource is renamed in later PR right?

File: Source/BasicNetworkService.swift [hunk 2]
```diff
@@ -46,6 +54,31 @@ public final class BasicNetworkService: NetworkService {
...
+        print(htmlText, response)
+     }, onError: { error in
+        // Handle errors
+     })
+     ```
+     
+     - parameter queue: The DispatchQueue to execute the completion and error block on.
+     - parameter resource: The resource you want to fetch.
+     - parameter onCompletionWithResponse: Callback which gets called when fetching and tranforming into model succeeds.
+     - parameter onError: Callback which gets called when fetching or tranforming fails.
```
Comment (IsabellSchmelzeisen) [inline Source/BasicNetworkService.swift]: tranforming -> transforming

File: Source/ModifyRequestNetworkService.swift [hunk 3]
```diff
@@ -46,6 +54,31 @@ public final class ModifyRequestNetworkService: NetworkService {
...
+    /**
+     Fetches a resource asynchronously from remote location. Execution of the requests starts immediately.
+     Execution happens on no specific queue. It dependes on the network access which queue is used.
+     Once execution is finished either the completion block or the error block gets called.
+     You can decide on which queue these blocks get called.
+     
+     **Example**:
+     ```swift
+     let networkService: NetworkService = //
+     let resource: Ressource<String> = //
```
Comment (IsabellSchmelzeisen) [inline Source/ModifyRequestNetworkService.swift]: Ressource -> Resource

File: Source/ModifyRequestNetworkService.swift [hunk 4]
```diff
@@ -46,6 +54,31 @@ public final class ModifyRequestNetworkService: NetworkService {
...
+     networkService.request(queue: .main, resource: resource, onCompletionWithResponse: { htmlText, response in
+        print(htmlText, response)
+     }, onError: { error in
+        // Handle errors
+     })
+     ```
+     
+     - parameter queue: The DispatchQueue to execute the completion and error block on.
+     - parameter resource: The resource you want to fetch.
+     - parameter onCompletionWithResponse: Callback which gets called when fetching and tranforming into model succeeds.
```
Comment (IsabellSchmelzeisen) [inline Source/ModifyRequestNetworkService.swift]: tranforming -> transforming

File: Source/NetworkError.swift [hunk 5]
```diff
@@ -23,17 +23,22 @@
...
 import Foundation
 
-/**
- `NetworkError` provides a collection of error types which can occur during execution.
- */
+/// `NetworkError` provides a collection of error types which can occur during execution.
 public enum NetworkError: Error {
+    /// The error is unkonw
     case unknownError
+    /// The request was canelled before it finished
```
Comment (IsabellSchmelzeisen) [inline Source/NetworkError.swift]: canelled -> cancelled

File: Source/NetworkService.swift [hunk 6]
```diff
@@ -26,17 +26,35 @@ import Dispatch
...
-     Fetches a resource asynchronously from remote location.
+     Fetches a resource asynchronously from remote location. Execution of the requests starts immediately.
+     Execution happens on no specific queue. It dependes on the network access which queue is used.
+     Once execution is finished either the completion block or the error block gets called.
+     You can decide on which queue these blocks get called.
+     
+     **Example**:
+     ```swift
+     let networkService: NetworkService = //
+     let resource: Ressource<String> = //
```
Comment (IsabellSchmelzeisen) [inline Source/NetworkService.swift]: Resource

File: Source/NetworkServiceMock.swift [hunk 7]
```diff
@@ -56,25 +60,29 @@ public final class NetworkServiceMock: NetworkService {
...
-     Fetches a resource asynchronously from remote location. Completion and Error block will be called on the main thread.
+     Fetches a resource asynchronously from remote location. Execution of the requests starts immediately.
+     Execution happens on no specific queue. It dependes on the network access which queue is used.
+     Once execution is finished either the completion block or the error block gets called.
+     You can decide on which queue these blocks get called.
      
+     **Example**:
      ```swift
+     let networkService: NetworkService = //
+     let resource: Ressource<String> = //
```
Comment (IsabellSchmelzeisen) [inline Source/NetworkServiceMock.swift]: Resource

File: Source/NetworkServiceMock.swift [hunk 8]
```diff
@@ -56,25 +60,29 @@ public final class NetworkServiceMock: NetworkService {
...
      }, onError: { error in
-     //Handle errors
+        // Handle errors
      })
      ```
      
+     - parameter queue: The DispatchQueue to execute the completion and error block on.
      - parameter resource: The resource you want to fetch.
-     - parameter onCompletion: Callback which gets called when fetching and tranforming into model succeeds.
+     - parameter onCompletionWithResponse: Callback which gets called when fetching and tranforming into model succeeds.
```
Comment (IsabellSchmelzeisen) [inline Source/NetworkServiceMock.swift]: transforming

File: Source/NetworkTaskMock.swift [hunk 9]
```diff
@@ -23,27 +23,36 @@
...
+    }
+    
+    /// Creates an `NetworkTaskMock` instance
     public init() {}
-    public var state: NetworkTaskMockState?
     
+    /// State of the network taks. Can be used to assert.
+    public private(set) var state: State?
+    
+    /// Cancle the request. Sets state to cancled.
```
Comment (IsabellSchmelzeisen) [inline Source/NetworkTaskMock.swift]: Cancel

File: Source/RetryNetworkService.swift [hunk 10]
```diff
@@ -56,22 +58,26 @@ public final class RetryNetworkService: NetworkService {
...
-     Fetches a resource asynchronously from remote location.
+     Fetches a resource asynchronously from remote location. Execution of the requests starts immediately.
+     Execution happens on no specific queue. It dependes on the network access which queue is used.
+     Once execution is finished either the completion block or the error block gets called.
+     You can decide on which queue these blocks get called.
      
+     **Example**:
      ```swift
+     let networkService: NetworkService = //
+     let resource: Ressource<String> = //
```
Comment (IsabellSchmelzeisen) [inline Source/RetryNetworkService.swift]: Resource

File: Source/RetryNetworkService.swift [hunk 11]
```diff
@@ -56,22 +58,26 @@ public final class RetryNetworkService: NetworkService {
...
      }, onError: { error in
-     //Handle errors
+        // Handle errors
      })
      ```
      
+     - parameter queue: The DispatchQueue to execute the completion and error block on.
      - parameter resource: The resource you want to fetch.
-     - parameter onCompletion: Callback which gets called when fetching and tranforming into model succeeds.
+     - parameter onCompletionWithResponse: Callback which gets called when fetching and tranforming into model succeeds.
```
Comment (IsabellSchmelzeisen) [inline Source/RetryNetworkService.swift]: transforming

File: Source/BasicNetworkService.swift (hunk 1 above)
Comment (lightsprint09) [reply Source/BasicNetworkService.swift]: I will have a look

**Pull request merged**